userspace/libsinsp/container_docker_linux.cpp

```diff
@@ -0,0 +1,527 @@
+
+	const Json::Value *cpu = nullptr;
+	if(walk_down_json(*linux, &cpu, "resources", "cpu") && cpu->isObject())
+	{
+		set_numeric(*cpu, "shares", container->m_cpu_shares);
+		set_numeric(*cpu, "quota", container->m_cpu_quota);
+		set_numeric(*cpu, "period", container->m_cpu_period);
+	}
+
+	const Json::Value *privileged;
+	if(walk_down_json(*linux, &privileged, "security_context", "privileged") && privileged->isBool())
+	{
+		container->m_privileged = privileged->asBool();
+	}
+
+	return true;
+}
+
+uint32_t get_pod_sandbox_ip(const std::string& pod_sandbox_id)
+{
+	runtime::v1alpha2::PodSandboxStatusRequest req;
+	runtime::v1alpha2::PodSandboxStatusResponse resp;
+	req.set_pod_sandbox_id(pod_sandbox_id);
+	req.set_verbose(true);
+	grpc::ClientContext context;
+	auto deadline = std::chrono::system_clock::now() + std::chrono::milliseconds(cri_timeout);
+	context.set_deadline(deadline);
+	grpc::Status status = cri->PodSandboxStatus(&context, req, &resp);
+
+	if (!status.ok()) {
+		return 0;
+	}
+
+	const auto& pod_ip = resp.status().network().ip();
+	if (pod_ip.empty()) {
+		return 0;
+	}
+
+	uint32_t ip;
+	if (inet_pton(AF_INET, pod_ip.c_str(), &ip) == -1)
+	{
+		ASSERT(false);
+		return 0;
+	}
+	else
+	{
+		return ip;
+	}
+}
+}
+#endif
+
+std::string sinsp_container_engine_docker::m_api_version = "/v1.24";
+
+sinsp_container_engine_docker::sinsp_container_engine_docker()
+{
+#if defined(HAS_CAPTURE)
+	if(!m_curlm)
+	{
+		m_curl = curl_easy_init();
+		m_curlm = curl_multi_init();
+
+		if(m_curlm)
+		{
+			curl_multi_setopt(m_curlm, CURLMOPT_PIPELINING, CURLPIPE_HTTP1|CURLPIPE_MULTIPLEX);
+		}
+
+		if(m_curl)
+		{
+			auto docker_path = scap_get_host_root() + docker_unix_socket_path;
+			curl_easy_setopt(m_curl, CURLOPT_UNIX_SOCKET_PATH, docker_path.c_str());
+			curl_easy_setopt(m_curl, CURLOPT_HTTPGET, 1);
+			curl_easy_setopt(m_curl, CURLOPT_FOLLOWLOCATION, 1);
+			curl_easy_setopt(m_curl, CURLOPT_WRITEFUNCTION, docker_curl_write_callback);
+		}
+	}
+
+	if(!cri && !cri_unix_socket_path.empty())
+	{
+		auto cri_path = scap_get_host_root() + cri_unix_socket_path;
+		struct stat s;
+		if(stat(cri_path.c_str(), &s) == 0 && (s.st_mode & S_IFMT) == S_IFSOCK)
+		{
+			cri = runtime::v1alpha2::RuntimeService::NewStub(
+				grpc::CreateChannel("unix://" + cri_path, grpc::InsecureChannelCredentials()));
+		}
+	}
+#endif
+}
+
+void sinsp_container_engine_docker::cleanup()
+{
+#if defined(HAS_CAPTURE)
+	curl_easy_cleanup(m_curl);
+	m_curl = NULL;
+	curl_multi_cleanup(m_curlm);
+	m_curlm = NULL;
+
+	cri.reset(nullptr);
+#endif
+}
+
+void sinsp_container_engine_docker::set_cri_socket_path(const std::string& path)
+{
+#if !defined(CYGWING_AGENT) && defined(HAS_CAPTURE)
+	cri_unix_socket_path = path;
+#endif
+}
+
+void sinsp_container_engine_docker::set_cri_timeout(int64_t timeout_ms)
+{
+#if !defined(CYGWING_AGENT) && defined(HAS_CAPTURE)
+	cri_timeout = timeout_ms;
+#endif
+}
+
+std::string sinsp_container_engine_docker::build_request(const std::string &url)
```

Comment:
We define it in _win/_linux implementation files but call it from _common, so it would have to be declared in a header anyway. I don't think making a function from the method would buy us much.